File: 0046-permutations/0046-permutations.java
class Solution {
  public List<List<Integer>> permute(int[] nums) {
    List numsList = new ArrayList<Integer>();
    for (int n : nums) {
      numsList.add(n);
    }
    return permute(numsList);
  }
  
  public List<List<Integer>> permute(List<Integer> numsList) {
    List ans = new ArrayList();
    if (numsList.size() == 1) {
      ans.add(numsList);
    } else {
      for (int i = 0; i < numsList.size(); i++) {
        List copy = new ArrayList<Integer>(numsList);
        int extracted = (int) copy.get(i);
        copy.remove(i);
        List<List<Integer>> permuted = permute(copy);
        for (List<Integer> p : permuted) {
          List<Integer> combined = new ArrayList<Integer>();
          combined.add(extracted);
          combined.addAll(p);
          ans.add(combined);
        }
      }
    }
    
    return ans;
  }
}

File: 0046-permutations/0046-permutations.go
func permute(nums []int) [][]int {
  ans := [][]int{}
  if len(nums) == 1 {
    ans = append(ans, nums)
  } else {
    for i:=0;i<len(nums);i++ {
      clone, extracted := make([]int, len(nums)-1), 0
      for j:=0;j<len(nums);j++ {
        if j < i {
          clone[j] = nums[j]
        } else if j == i {
          extracted = nums[j]
        } else {
          clone[j-1] = nums[j]
        }
      }
      permuted := permute(clone)
      for _, p := range permuted {
        combined := []int{extracted}
        combined = append(combined, p...)
        ans = append(ans, combined)
      }
    }
  }
  return ans
}